pkg/skaffold/build/ko/build_test.go

```diff
@@ -0,0 +1,153 @@
+			ref:                 "gcr.io/project-id/test-app2:testTag",
+			imageID:             "testTag",
+			pushImages:          true,
+			importpath:          "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/ko",
+			imageNameFromConfig: "test-app2",
+		},
+		{
+			description:         "ko import path image name and push image",
+			ref:                 "gcr.io/project-id/example.com/myapp:myTag",
+			imageID:             "myTag",
+			pushImages:          true,
+			importpath:          "ko://example.com/myapp",
+			imageNameFromConfig: "ko://example.com/myapp",
+		},
+		{
+			description:         "workspace is not cwd",
+			ref:                 "gcr.io/project-id/example.com/test-app3:myTag",
+			imageID:             "imageID3",
+			pushImages:          false,
+			importpath:          "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			imageNameFromConfig: "test-app3",
+			workspace:           "../docker",
+		},
+		{
+			description:         "ko import path image name and workspace is not cwd",
+			ref:                 "gcr.io/project-id/example.com/test-app4:myTag",
+			imageID:             "imageID4",
+			pushImages:          false,
+			importpath:          "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			imageNameFromConfig: "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			workspace:           "../docker",
+		},
+		{
+			description:         "ko import path image name and workspace is not cwd and import path is subdirectory of cwd",
+			ref:                 "gcr.io/project-id/example.com/test-app5:myTag",
+			imageID:             "imageID5",
+			pushImages:          false,
+			importpath:          "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			imageNameFromConfig: "ko://github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/docker",
+			workspace:           "..",
+		},
+	}
+	for _, test := range tests {
+		testutil.Run(t, test.description, func(t *testutil.T) {
+			b := stubKoArtifactBuilder(test.ref, test.imageID, test.pushImages, test.importpath)
+
+			artifact := &latestV1.Artifact{
+				ArtifactType: latestV1.ArtifactType{
+					KoArtifact: &latestV1.KoArtifact{},
+				},
+				Dependencies: []*latestV1.ArtifactDependency{},
+				ImageName:    test.imageNameFromConfig,
+				Workspace:    test.workspace,
+			}
+
+			var outBuffer bytes.Buffer
+			gotImageID, err := b.Build(context.TODO(), &outBuffer, artifact, test.ref)
+			t.CheckNoError(err)
+			if gotImageID != test.imageID {
+				t.Errorf("got image ID %s, wanted %s", gotImageID, test.imageID)
+			}
+			imageNameOut := strings.TrimSuffix(outBuffer.String(), "\n")
+			if imageNameOut != test.ref {
+				t.Errorf("image name output was %q, wanted %q", imageNameOut, test.ref)
+			}
+		})
+	}
+}
+
+func stubKoArtifactBuilder(ref string, imageID string, pushImages bool, importpath string) *Builder {
```

Comment:
Is this better as `subKoArtifactPublisher`?  The builder isn't stubbed.

Reply:
It returns a `Builder`, so how about `builderWithStubPublish()`?

The `docker.LocalDaemon` is also fake though, so that name doesn't really indicate that both the `publishImages()` function and the `localDocker` field are fakes.